Global Trigger data types & jets pattern files

thesps opened the pull request:
I've added `gt_datatypes.h` with the data formats from the Global Trigger interface document for jets, sums, taus, electrons and photons, as well as conversions for the common fields (pT, eta and phi) from the Correlator internal coordinates.

Then, I've added a producer to create pattern files from the GT-formatted jets using the DemonstratorTools stuff.

At the moment, I've tested that it compiles and I can produce a pattern file that looks sensible when decoded by hand compared to the jets in the ntuple. (With [this config file](https://gist.github.com/thesps/8616ea1a89f42613de843cc7ca387d11)).

The corresponding correlator-common MR [is here](https://gitlab.cern.ch/cms-cactus/phase2/firmware/correlator-common/-/merge_requests/62).

thesps committed:
Implement DemonstratorTools producer for l1t::PFJet pattern files. Implement gt_datatypes.h for Global Trigger data formats, and conversions from Correlator internal coordinates

thesps committed:
Make more parameters configurable, pad output link with valid but empty jets up to constant nJets (12 by default)

gpetruc commented on L1Trigger/Phase2L1ParticleFlow/plugins/L1CTJetFileWriter.cc:
```diff
@@ -0,0 +1,109 @@
+  fileWriterOutputToGT_.flush();
+}
+
+std::array<std::vector<ap_uint<64>>, 1> L1CTJetFileWriter::encodeJets(const edm::View<l1t::PFJet>& jets) {
```
why do you need the `std::array<X,1>` and it can't be just `std::vector<ap_uint<64>>` ?
eventually it seems to me it's only accessed above at line 72 with `.at(0)` and so the `std::array` interface itself is never used

gpetruc commented on L1Trigger/Phase2L1ParticleFlow/python/l1ctJetFileWriter_cff.py:
```diff
@@ -0,0 +1,11 @@
+import FWCore.ParameterSet.Config as cms
```
should probably be a `_cfi` instead of a `_cff` since it's defining a single module directly

gpetruc commented on L1Trigger/Phase2L1ParticleFlow/plugins/L1CTJetFileWriter.cc:
```diff
@@ -0,0 +1,109 @@
+  //The following says we do not know what parameters are allowed so do no validation
+  // Please change this to state exactly what you do use, even if it is no parameters
+  edm::ParameterSetDescription desc;
+  desc.setUnknown();
```
perhaps since the method is defined one could also put here the parameters
```cpp
desc.add<edm::InputTag>("jets");
desc.add<std::string>("outputFilename");
desc.add<uint32_t>("nJets", 12);
desc.add<uint32_t>("nFramesPerBX", 9);
desc.add<uint32_t>("TMUX", 6);
desc.add<uint32_t>("maxLinesPerFile", 1024);
desc.add<std::string>("format","EMP");
```

gpetruc reviewed (changes requested):
Few code improvements suggested + sync with https://gitlab.cern.ch/cms-cactus/phase2/firmware/correlator-common/-/merge_requests/62 after updating the code there.

thesps commented on L1Trigger/Phase2L1ParticleFlow/plugins/L1CTJetFileWriter.cc:
```diff
@@ -0,0 +1,109 @@
+      nJets(iConfig.getParameter<unsigned>("nJets")),
+      kFramesPerBX(iConfig.getParameter<unsigned>("nFramesPerBX")),
+      kCTL2BoardTMUX(iConfig.getParameter<unsigned>("TMUX")),
+      kGapLengthOutput(kFramesPerBX - 2 * nJets),
```
@cerminar pointed out this should instead be be `kCTL2BoardTMUX * kFramesPerBX - 2 * nJets`

thesps committed:
Sync GT dataformats with correlator-common https://gitlab.cern.ch/cms-cactus/phase2/firmware/correlator-common/-/merge_requests/62

thesps committed:
Remove std::array interface from encode; mv l1ctJetFileWriter cff->cfi; set the default descriptions; add missing jet sort

thesps commented:
I've done the sync with [the correlator-common MR](https://gitlab.cern.ch/cms-cactus/phase2/firmware/correlator-common/-/merge_requests/62), and addressed the other comments above. 

I also added a missing sort to the jets. I didn't notice it before since the jets were being sorted somewhere else in the flow of making ntuples, but the collection returned by the emulator was not sorted. Since the firmware does do a sort, I think it makes sense to be a part of the emulation directly. Plus, it's required to do the sort somewhere for the pattern files to get the correct ordering.

gpetruc commented on L1Trigger/Phase2L1ParticleFlow/plugins/L1CTJetFileWriter.cc:
```diff
@@ -0,0 +1,114 @@
+  size_t kCTL2BoardTMUX; 
+  size_t kGapLengthOutput; 
+  size_t kMaxLinesPerFile; 
+  std::map<l1t::demo::LinkId, std::pair<l1t::demo::ChannelSpec, std::vector<size_t>>> kChannelSpecsOutputToGT; 
```
per CMS coding style, all these names should have a trailing underscore character (`nJets_`, ...)

gpetruc reviewed (changes requested):
also, it would need a scram b code-format (applying it only to the files in this branch ... unfortunately the cms-l1t-offline has accumulated other formatting issues but I think it's not up to us to fix them)

thesps committed:
Naming and style compliance

thesps commented:
I've updated after applying the code formatting

thesps committed:
Move the jet sort out of the emulator (needs to happen after corrections)

thesps commented:
I've moved the sort out of the emulator since it caused mismatch wrt firmware - the sort happens after the JECs.
